jerry-core/api/jerry.c

```diff
@@ -3088,6 +3089,75 @@ jerry_typedarray_find_by_type (jerry_typedarray_type_t type_name, /**< type of t
 
 #endif /* !CONFIG_DISABLE_ES2015_TYPEDARRAY_BUILTIN */
 
+/**
+ * Creates a RegExp object with the given pattern and flags.
+ *
+ * @return value of the Constructed RegExp object.
+ */
+jerry_value_t
+jerry_create_regexp (const jerry_char_t *pattern, /**< RegExp pattern */
+                    jerry_length_t pattern_length, /**< Length of the pattern */
+                    jerry_regexp_flags_t flags) /**< Optional RegExp flags */
+{
+  jerry_assert_api_available ();
+
+#ifndef CONFIG_DISABLE_REGEXP_BUILTIN
+  if (!lit_is_valid_utf8_string ((lit_utf8_byte_t *) pattern, (lit_utf8_size_t) pattern_length))
+  {
+    return jerry_throw (ecma_raise_common_error ("Input must be a valid utf8 string"));
+  }
+
+  ecma_string_t *ecma_pattern = ecma_new_ecma_string_from_utf8 ((lit_utf8_byte_t *) pattern,
+                                                                      (lit_utf8_size_t) pattern_length);
+
+  ecma_value_t str_val = ecma_make_string_value (ecma_pattern);
+
+  jerry_value_t ret_val;
+
+  if (flags > 0)
+  {
+    ecma_value_t flag_val = ECMA_VALUE_UNDEFINED;
+    jerry_char_t flag_str[3];
+    unsigned char *str_ptr = flag_str;
+
+    if (flags & JERRY_RE_FLAG_GLOBAL)
+    {
+      *str_ptr++ = 'g';
+    }
+
+    if (flags & JERRY_RE_FLAG_MULTILINE)
+    {
+      *str_ptr++ = 'm';
+    }
+
+    if (flags & JERRY_RE_FLAG_IGNORE_CASE)
+    {
+      *str_ptr++ = 'i';
+    }
+
+    flag_val = jerry_create_string_sz (flag_str, (jerry_size_t) (str_ptr - flag_str));
```

Comment:
That's true, however I don't know if we should modify the ecma_op_create_regexp_object function just for the sake of creating an API function.
Could you explain a bit more what do you mean by creating a new variant of that function? A new function with a different name, with similar functionality ? Or modifying the current function? I think the latter one wouldn't be the best solution, since the internal ecma function is used way more often than the jerry_api function would be, and the performance overhead is better to have in the API function IMHO.
However, I would like to hear your opinion on that as well.

Reply:
I've taken a new look at the code. `ecma_op_create_regexp_object` seems only to be called once, from `ecma_builtin_regexp_dispatch_construct`. Thus, it doesn't need a new variant, actually, but the flag parsing can be factored out from it to the only call site. That is, `ecma_op_create_regexp_object` could be:

```c
ecma_value_t
ecma_op_create_regexp_object (ecma_string_t *pattern_p, /**< input pattern */
                              uint16_t flags) /**< flags */
```

And `ecma_builtin_regexp_dispatch_construct` could contain the following logic:

```c
  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
  uint16_t flags = 0;

  if (flags_str_p != NULL)
  {
    ECMA_TRY_CATCH (empty, re_parse_regexp_flags (flags_str_p, &flags), ret_value);
    ECMA_FINALIZE (empty);

    if (!ecma_is_value_empty (ret_value))
    {
      return ret_value;
    }
  }
```

(This might need some adaptation / simplification. E.g., might not need to return.)

And in this case, code here (some lines below) could simply be `ret_val = ecma_op_create_regexp_object (ecma_pattern, (uint16_t) flags);`.

Reply:
Thank you for the explanation, much appreciated.